core/llm_google.go

```diff
@@ -164,91 +122,193 @@ func (c *GenaiClient) SendQuery(ctx context.Context, history []ModelMessage, too
+				err = fmt.Errorf("unexpected genai part type %T", part)
+				return content, toolCalls, tokenUsage, err
 			}
 		}
 	}
 
+	return content, toolCalls, tokenUsage, nil
+}
+
+func (c *GenaiClient) SendQuery(ctx context.Context, history []ModelMessage, tools []LLMTool) (_ *LLMResponse, rerr error) {
+	// setup tracing & telemetry
+	ctx, span := Tracer(ctx).Start(ctx, "LLM query", telemetry.Reveal(), trace.WithAttributes(
+		attribute.String(telemetry.UIActorEmojiAttr, "🤖"),
+		attribute.String(telemetry.UIMessageAttr, "received"),
+	))
+	defer telemetry.End(span, func() error { return rerr })
+
+	stdio := telemetry.SpanStdio(ctx, InstrumentationLibrary,
+		log.String(telemetry.ContentTypeAttr, "text/markdown"))
+	defer stdio.Close()
+
+	m := telemetry.Meter(ctx, InstrumentationLibrary)
+	attrs := []attribute.KeyValue{
+		attribute.String(telemetry.MetricsTraceIDAttr, span.SpanContext().TraceID().String()),
+		attribute.String(telemetry.MetricsSpanIDAttr, span.SpanContext().SpanID().String()),
+		attribute.String("model", c.endpoint.Model),
+		attribute.String("provider", string(c.endpoint.Provider)),
+	}
+
+	inputTokens, err := m.Int64Gauge(telemetry.LLMInputTokens)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get inputTokens gauge: %w", err)
+	}
+
+	inputTokensCacheReads, err := m.Int64Gauge(telemetry.LLMInputTokensCacheReads)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get inputTokensCacheReads gauge: %w", err)
+	}
+
+	outputTokens, err := m.Int64Gauge(telemetry.LLMOutputTokens)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get outputTokens gauge: %w", err)
+	}
+
+	// setup model
+	model := c.client.GenerativeModel(c.endpoint.Model)
+
+	// convert tools
+	genaiTools := c.convertToolsToGenai(tools)
+	// Gemini expects non empty FunctionDeclarations when tools are provided
+	if len(genaiTools) > 0 {
+		model.Tools = genaiTools
+	}
+
+	// convert history
+	if len(history) == 0 {
+		return nil, fmt.Errorf("genai history cannot be empty")
+	}
+	genaiHistory, systemInstruction, err := c.convertHistoryToGenai(history)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert history: %w", err)
+	}
+	model.SystemInstruction = systemInstruction
+
+	if len(genaiHistory) == 0 {
+		return nil, fmt.Errorf("no user prompt")
+	}
+
+	userMessage := genaiHistory[len(genaiHistory)-1]
+	chatHistoryForGenai := genaiHistory[:len(genaiHistory)-1]
+
+	chat := model.StartChat()
+	chat.History = chatHistoryForGenai
+
+	stream := chat.SendMessageStream(ctx, userMessage.Parts...)
+
+	// log text chunk to the telemetry span's output
+	textHandler := func(ctx context.Context, text string) error {
+		_, err := fmt.Fprint(stdio.Stdout, text)
+		return err
+	}
+
+	// records token usage metrics and updates the final summary struct based on metadata from the stream.
+	tokenHandler := func(ctx context.Context, usageMeta *genai.UsageMetadata, usageSummary *LLMTokenUsage) error {
+		candidateTokens := int64(usageMeta.CandidatesTokenCount)
```

Comment:
missing `s` in `candidate` in `candidatesTokens`. Also probably better to call it `candidatesTokenCount` like the usageMeta field. Same for the other fields.